### PR TITLE
Tag every character

### PR DESCRIPTION
**Pull request by RichartE**
This pull request adds the following functionality:
- Adds html mark tags around every character in the wikipedia body-content
- Exposes a clean text that can be compared to find index matches
- Exposes API to mark index ranges
- Exposes API to remove all marks

This pull request also reorganizes the code a bit. There is now a root.js file where the methods are imported to reduce the code complexity. The goal is to have each logical area be in its own file and the root just connecting them all as needed.

### CODE REVIEW

**Comment by RichartE** [comment]
you can import the authoritative html text from the tagEveryChar.js file. To highlight characters you need to use the markPage function (from the markPage.js file) that takes a start and end (inclusive) index matching the text. To clear the highlights use the removeMarks funciton from markPage.js file.

File: wikichange/src/content/content.js [hunk 1]
```diff
@@ -360,9 +360,11 @@ const highlightRevisionBetweenRevisionIds = async (title, curRevisionId, oldRevi
  */
 const highlight = async (revisionId, oldRevisionId) => {
     const arr = await fetchChangeWithHTML(oldRevisionId, revisionId);
+    console.groupCollapsed('highlight-messages');
     arr.forEach((element) => {
         highlightContentUsingNodes(element, "#AFE1AF");
     });
+    console.groupEnd();
```
Comment (sukritsangvong) [inline wikichange/src/content/content.js]: do we need to keep these console.log()s?

**Review by sukritsangvong — CHANGES_REQUESTED** [review]
I tried opening the United States page (https://en.wikipedia.org/wiki/United_States) and the page totally crashed on me. The Pasta page was also pretty so highlight the first sentence and remove it. Can you please make sure that this works on the United States page on your machine? If it does, then it's probably fine. My machine is kinda slow today. If it doesn't, you probably have to optimize something to make it faster.

File: wikichange/src/content/content.js (hunk 1 above)
Comment (LuisaE) [reply wikichange/src/content/content.js]: Remove console.logs, it will further slow down the page

**Comment by LuisaE** [comment]
Great! We just need to make sure this will not crash, so maybe merge without actually calling it? So I can start my part testing locally.

File: wikichange/src/content/root.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { text } from "./tagEveryWord";
+import { markPage, removeMarks } from "./markPage";
+import "./content";
+
+console.info(text.substring(0, 21));
```
Comment (LuisaE) [inline wikichange/src/content/root.js]: Let's move this code to the readme, as it is test code and not actually "prod ready"

Comment (LuisaE) [reply wikichange/src/content/root.js]: Eg: ### How to tag every word

File: wikichange/src/content/tagEveryWord.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ * @return {string} Returns a the concatenated text where each character index matches the mark id 
+ */
+const tagEveryWord = () => {
+    const bodyContent = document.getElementById('bodyContent');
+    const treeWalker = document.createTreeWalker(
+        bodyContent,
+        NodeFilter.SHOW_TEXT,
+        (node) => {
+            if (node.parentElement.tagName === 'STYLE') {
+                return NodeFilter.FILTER_REJECT;
```
Comment (LuisaE) [inline wikichange/src/content/tagEveryWord.js:12]: Why do we need this and what do it do?

File: wikichange/assets/content.css [hunk 4]
```diff
@@ -0,0 +1,4 @@
+mark{
```
Comment (LuisaE) [inline wikichange/assets/content.css:1]: Thanks for creating a css file, maybe when this is merged @ducndh can add all the CSS here?

**Comment by sukritsangvong** [comment]
Tested again and it crashed on the United States page the first couple of times then able to load on 3rd or 4th try. It froze after the first couple of words were highlighted in yellow and the loader was spinning. (the graph not showing up). This might be a combination of a couple of things. will take a deeper look into this.

image

File: wikichange/src/content/compareRevisionService.js [hunk 5]
```diff
@@ -15,7 +15,9 @@ const fetchChangeWithHTML = async (startID, endID) => {
     const divsWithNoInsOuts = [...allDivs].filter((curDiv) => {
         return curDiv.querySelectorAll("ins.diffchange.diffchange-inline,del.diffchange.diffchange-inline").length == 0;
     });
+    console.groupCollapsed('divs-with-no-in-outs');
     divsWithNoInsOuts.forEach((x) => console.log(x.innerHTML));
```
Comment (sukritsangvong) [inline wikichange/src/content/compareRevisionService.js]: I think this was left over from my pr, but can u please remove this line.

**Comment by sukritsangvong** [comment]
Checked with the main, and the graph loading was slow, but it wasn't slow enough to cause the page to crash. There is probably something in this pr that is slowing this down to crossing the threshold.

**Review by sukritsangvong — COMMENTED** [review]
Everything looks good. Will have Luisa be the one to approve this since she will have to be working with this (want her to at least see them before merging). @LuisaE

**Review by LuisaE — APPROVED** [review]
Thank you, looks great!

**Comment by LuisaE** [comment]
> Everything looks good. Will have Luisa be the one to approve this since she will have to be working with this (want her to at least see them before merging). @LuisaE

Approved!

**Review by sukritsangvong — APPROVED** [review]

**Pull request merged**